File: go/MinDepthBinaryTree/MinDepthBinaryTree.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time -> 0(N) Space -> 0(log(N))
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return findDepth(root, 1)
}

func findDepth(root *TreeNode, depth int) int {
	if root.Left == nil && root.Right == nil {
		return depth
	}

	newDepth := depth + 1
	leftDepth := math.MaxInt
	rightDepth := math.MaxInt
	if root.Left != nil {
		leftDepth = findDepth(root.Left, newDepth)
	}
	if root.Right != nil {
		rightDepth = findDepth(root.Right, newDepth)
	}

	return min(leftDepth, rightDepth)
}

func main() {
	left := &TreeNode{Val: 9}
	// right := &TreeNode{Val: 20}
	root := &TreeNode{Val: 3, Left: left}
	fmt.Println(minDepth(root))
}

File: python/minDepthBinaryTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


# Time -> 0(N) Space -> 0(log(N))
def minDepth(root) -> int:
    if root is None:
        return 0
    return findDepth(root, 1)


def findDepth(root, depth) -> int:
    if root.left is None and root.right is None:
        return depth

    newDepth = depth + 1
    leftDepth = findDepth(root.left, newDepth) if root.left is not None else -1
    rightDepth = findDepth(root.right, newDepth) if root.right is not None else -1

    if leftDepth == -1 or rightDepth == -1:
        if leftDepth == -1:
            return rightDepth
        return leftDepth
    return min(leftDepth, rightDepth)


right = TreeNode(val=20)
left = TreeNode(val=9)
root = TreeNode(val=3, left=left, right=right)
print(minDepth(root))
